pepper-apis/src/main/java/org/broadinstitute/ddp/route/PatchFormAnswersRoute.java

```diff
@@ -152,51 +149,63 @@ public Object handle(Request request, Response response) {
-                String msg = "Activity instance with GUID " + instanceGuid + " is read-only, cannot submit answer(s) for it";
-                LOG.info(msg);
-                throw ResponseUtil.haltError(response, 422, new ApiError(ErrorCodes.ACTIVITY_INSTANCE_IS_READONLY, msg));
+            ActivityDefStore activityStore = ActivityDefStore.getInstance();
+            FormActivityDef activityDef = ActivityInstanceUtil.getActivityDef(handle, activityStore, instanceDto, studyGuid);
+
+            ActivityInstanceDto parentInstanceDto = null;
+            FormActivityDef parentActivityDef = null;
+            String parentInstanceGuid = instanceDto.getParentInstanceGuid();
+            if (parentInstanceGuid != null) {
+                parentInstanceDto = instanceSummary
+                        .getActivityInstanceByGuid(parentInstanceGuid)
+                        .orElseThrow(() -> new DDPException("Could not find parent instance " + parentInstanceGuid));
+                parentActivityDef = ActivityInstanceUtil.getActivityDef(handle, activityStore, parentInstanceDto, studyGuid);
+            }
+
+            if (!isStudyAdmin) {
+                if (parentInstanceDto != null && ActivityInstanceUtil.isInstanceReadOnly(parentActivityDef, parentInstanceDto)) {
+                    String msg = "Parent activity instance with GUID " + parentInstanceDto.getGuid()
+                            + " is read-only, cannot submit answer(s) for child instance " + instanceGuid;
+                    LOG.info(msg);
+                    throw ResponseUtil.haltError(response, 422, new ApiError(ErrorCodes.ACTIVITY_INSTANCE_IS_READONLY, msg));
+                }
+                if (ActivityInstanceUtil.isInstanceReadOnly(activityDef, instanceDto)) {
+                    String msg = "Activity instance with GUID " + instanceGuid + " is read-only, cannot submit answer(s) for it";
+                    LOG.info(msg);
+                    throw ResponseUtil.haltError(response, 422, new ApiError(ErrorCodes.ACTIVITY_INSTANCE_IS_READONLY, msg));
+                }
             }
 
             var answerDao = new AnswerCachedDao(handle);
             var questionDao = new QuestionCachedDao(handle);
             var jdbiQuestion = questionDao.getJdbiQuestion();
 
-            LanguageDto preferredUserLanguage = RouteUtil.getUserLanguage(request);
-            String isoLanguageCode = preferredUserLanguage.getIsoCode();
-            Long languageCodeId = preferredUserLanguage.getId();
+            LanguageDto preferredUserLangDto = RouteUtil.getUserLanguage(request);
+
+            User operatorUser = instanceSummary.getParticipantUser();
+            if (!participantGuid.equals(operatorGuid)) {
+                operatorUser = handle.attach(UserDao.class).findUserByGuid(operatorGuid)
+                        .orElseThrow(() -> new DDPException("Could not find operator with guid " + operatorGuid));
+            }
```

Comment:
Small optimization: operator is almost always the participant, so we start with that and only do a query if operator is different. Also putting this down here so we only do this query if it passed everything else above.